compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1565,16 +1566,138 @@ fn linker_with_args<'a, B: ArchiveBuilder<'a>>(
     let cmd = &mut *codegen_results.linker_info.to_linker(base_cmd, &sess, flavor);
     let link_output_kind = link_output_kind(sess, crate_type);
 
-    // NO-OPT-OUT, OBJECT-FILES-MAYBE, CUSTOMIZATION-POINT
+    // ------------------------------ Early order-dependent options -------------------------------
+
+    // Avoid linking to dynamic libraries unless they satisfy some undefined symbols
+    // at the point at which they are specified on the command line.
+    // Must be passed before any (dynamic) libraries to have effect on them.
+    cmd.add_as_needed();
+
+    // If we're building something like a dynamic library then some platforms
+    // need to make sure that all symbols are exported correctly from the
+    // dynamic library.
+    // Must be passed before any libraries to prevent the symbols to export from being thrown away,
+    // at least on some platforms (e.g. windows-gnu).
+    cmd.export_symbols(tmpdir, crate_type);
+
+    // Can be used for adding custom CRT objects or overriding order-dependent options above.
+    // FIXME: In practice built-in target specs use this for arbitrary order-independent options,
+    // introduce a target spec option for order-independent linker options and migrate built-in
+    // specs to it.
     add_pre_link_args(cmd, sess, flavor);
 
-    // NO-OPT-OUT, OBJECT-FILES-NO
+    // ---------------------- Object code and libraries, order-dependent  -------------------------
+
+    // Pre-link CRT objects.
+    add_pre_link_objects(cmd, sess, link_output_kind, crt_objects_fallback);
```

Comment:
AFAICT in the past we did not `export_symbols` from the CRT objects nor the sanitizer libraries. Are there bugs that are fixed by this re-ordering? And if so, can they be linked in the commit message, or at least this change called out if there isn't an issue?

Reply:
Moving `export_symbols` to before `add_pre_link_objects` doesn't fix any issues, it's just for gathering objects/libraries into one group and options to another.